### PR TITLE
Single-prefix+bitfield token encoding

### PR DESCRIPTION
**Pull request by A60AB5450353F40E**
Implement #28

Also found some nits in introspection opcode definitions and test vectors

### CODE REVIEW

**Comment by bitjson** [comment]
Thanks for the PR @A60AB5450353F40E! Still working on reviewing (by implementing in Libauth and updating those test vectors) – should have solid thoughts here in a few days. 👍

File: readme.md [hunk 1]
```diff
@@ -110,23 +110,31 @@ Every transaction output can optionally have a [`token category`](#token-categor
...
-   1. `0x00` – no NFT – the prefix **does not encode a non-fungible token**. Skip to `ft_amount`.
-   2. `0x01` – the **`mutable` capability** – the encoded non-fungible token is a **mutable token**.
-   3. `0x02` – the **`minting` capability** – the encoded non-fungible token is a **minting token**.
-   4. `0x11` – no capability – the encoded non-fungible token is an **immutable token**.
-3. `commitment_length` (if an NFT is present) – A **commitment length** (encoded in `CompactSize` format<sup>1</sup>).
-4. `commitment` – If `commitment_length` is non-zero, a **token commitment** byte string of `commitment_length` is required.
-5. `ft_amount` – a **token amount** (encoded as in `CompactSize` format) with a minimum value of `0` (`0x00`) and a maximum value equal to the maximum VM number, `9223372036854775807` (`0xffffffffffffff7f`).
+2. `<token_format | nft_capability>` - A byte encoding two half-byte (4-bit) fields
+    1. `<token_format> - A bitfield indicating the token payload that follows, defined at the higher half of the serialized byte, to be deserialized using `token_format = serialized_byte & 0xf0`.
```
Comment (zander) [inline readme.md:125]: nit-pick. Serialization is to turn something into bytes so using **serialized byte** is non-sensical. Its just a byte. Bitfield is indeed the best name that all programmers should understand.

File: readme.md [hunk 2]
```diff
@@ -110,23 +110,31 @@ Every transaction output can optionally have a [`token category`](#token-categor
...
+        1. `0x10` (`b00010000`) - Fungible tokens
+        2. `0x40` (`b01000000`) - Non-fungible token
+        3. `0x50` (`b01010000`) - Both fungible tokens and a non-fungible token
+        4. `0x60` (`b01100000`) - Non-fungible token with a commitment
+        5. `0x70` (`b01110000`) - Both fungible tokens and a non-fungible with a commitment
+    2. `<nft_capability>` – A byte indicating the capability of a non-fungible token, defined at the lower half of the serialized byte, to be deserialized using `token_format = serialized_byte & 0x0f`.
```
Comment (zander) [inline readme.md:131]: Starting with "A byte" is confusing as you are simply continuing the bitfield from above.

Or, in other words, you are listing the bit-field fields and you are at the same time listing the way it is encoded and that is confusing.

I suggest simply stating it is a bitfield and listing its items. This allows you to list less (for instance commitment is a single bit, but you list multiple combinations above, actually making it more confusing).

For an example, check the docs on https://doc.qt.io/qt-6/qflags.html#details specifically the 'example' which simply lists each bitfield.

File: readme.md (hunk 1 above)
Comment (A60AB5450353F40E) [reply readme.md]: How's this: https://github.com/bitjson/cashtokens/pull/33/commits/10707fe99aacc8ec69636251eeb189fbcefa109a

File: readme.md (hunk 2 above)
Comment (A60AB5450353F40E) [reply readme.md]: I liked Jason's style of listing only allowed combinations (we don't allow no tokens, or a commitment without an NFT) so thought to keep it consistent with old ver.

Comment (zander) [reply readme.md]: The allowed combinations is listed a couple of paragraphs below, though.

In a technical documentation is helps to just stick to the technical docs without fluff. These are flags in a bitfield and each can be documented individually.  
That way any technical person can read it faster and without fault and doesn't have to read the fluffy descriptions in order to find out that there is an undocumented "commitment" flag.

Comment (A60AB5450353F40E) [reply readme.md]: How's this: https://github.com/bitjson/cashtokens/pull/33/commits/deb25d64a2f4a7cbca88ad31aa6c87353520afe6

Comment (zander) [reply readme.md]: looks great!

**Comment by A60AB5450353F40E** [comment]
@bitjson not important, it just made sense to put them in the same order the related fields will appear

**Comment by bitjson** [comment]
Merged in https://github.com/bitjson/cashtokens/commit/e61ee768f9a27937b5b9caa803133eb448ff6bb8, thanks!

**Pull request closed without merging**